### PR TITLE
Validate that CodeScope is the ElementId of an existing element

### PR DESCRIPTION
**Pull request by kabentley**
The value of `CodeScope` *must* be a valid ElementId, but we did not enforce that. Throw an exception if the supplied value is illegal.

On insert and update, the value supplied from JavaScript may be a `FederationgGuid`. If so, look up the element and use its ElementId.

### CODE REVIEW

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp [hunk 1]
```diff
@@ -566,7 +568,26 @@ DgnCode DgnCode::FromJson(BeJsConst value, DgnDbCR db) {
...
+    // the code scope must refer to an existing element. The json may have an ElementId or FederationGuid
+    auto scopeJson = value[json_scope()];
+    auto scopeStr = scopeJson.asCString();
+
+    // Note: this is necessary since GetId64 parses strings that start with number successfully and can get confused with Guids.
+    // ElementIds in JSON must start with "0x"
+    auto validId = scopeStr && (scopeStr[0] == '0' && (scopeStr[1] == 'X' || scopeStr[1] == 'x'));
+
+    auto scopeId = scopeJson.GetId64<DgnElementId>();
+    if (!validId || !db.Elements().ElementExists(scopeId)) {
```
Comment (pmconne) [inline iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:581]: Did you decide not to special-case "0x1" after all?

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp [hunk 2]
```diff
@@ -566,7 +568,26 @@ DgnCode DgnCode::FromJson(BeJsConst value, DgnDbCR db) {
...
-    val.m_scope = value[json_scope()].asString();
-    val.m_value = value[json_value()].asString();
+
+    // the code scope must refer to an existing element. The json may have an ElementId or FederationGuid
+    auto scopeJson = value[json_scope()];
+    auto scopeStr = scopeJson.asCString();
+
+    // Note: this is necessary since GetId64 parses strings that start with number successfully and can get confused with Guids.
+    // ElementIds in JSON must start with "0x"
+    auto validId = scopeStr && (scopeStr[0] == '0' && (scopeStr[1] == 'X' || scopeStr[1] == 'x'));
```
Comment (pmconne) [inline iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:578]: You are permitting uppercase X in Ids? We forbid it elsewhere.

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp [hunk 3]
```diff
@@ -566,7 +568,26 @@ DgnCode DgnCode::FromJson(BeJsConst value, DgnDbCR db) {
...
+
+    auto scopeId = scopeJson.GetId64<DgnElementId>();
+    if (!validId || !db.Elements().ElementExists(scopeId)) {
+        BeGuid scopeFederationGuid;
+        DgnElementCPtr scopeEl;
+        if (BentleyStatus::SUCCESS == scopeFederationGuid.FromString(scopeJson.asCString()))
+            scopeEl = db.Elements().QueryElementByFederationGuid(scopeFederationGuid);
```
Comment (pmconne) [inline iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:585]: You expressed some concerns about additional overhead imposed by verifying scope element exists. Are you ok with:
- Querying for and loading entire element vs just its federation Guid; and
- If caller supplies a valid but non-existent element Id, attempting fruitlessly to look up an element that has that Id as its federation Guid?

**Review by pmconne — COMMENTED** [review]
Are you concerned about any iModels that may exist with invalid code scopes beginning to throw exceptions in certain contexts making them unusable?
Would it be safer to only throw from `DgnCode::FromJson` during insert and update?

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp (hunk 2 above)
Comment (kabentley) [reply iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:578]: That code comes from BeJsConst::GetUInt64. Should we not allow it there?

**Comment by kabentley** [comment]
> Are you concerned about any iModels that may exist with invalid code scopes beginning to throw exceptions in certain contexts making them unusable? Would it be safer to only throw from `DgnCode::FromJson` during insert and update?

Currently `DgnCode::FromJson` is only called from insert/update. But this logic should apply to any new use case of it too.

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp (hunk 2 above)
Comment (pmconne) [reply iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:578]: Meh, I don't know. BeInt64Id uses BeStringUtilities::HasHexPrefix which allows uppercase too.
TypeScript only permits lowercase.

Comment (kabentley) [reply iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:578]: that's why i used `scopeId.ToHexStr()` so that it will always be "correct" in the iModel.

**Comment by pmconne** [comment]
> No, `DgnCode::FromJson` is not called from GetElement/GetModel.

[Yes, it is](https://github.com/iTwin/imodel-native/blob/137a73e655d7a1336f181a89c3f93a216b3143b7/iModelJsNodeAddon/JsInteropDgnDb.cpp#L317).

[This caller](https://github.com/iTwin/itwinjs-core/blob/b99c65cdddf8b8e4975cff218c42b813864ab48e/core/backend/src/IModelDb.ts#L1578) expects any error thrown by the addon to have an `errorNumber`.

**Comment by kabentley** [comment]
> > No, `DgnCode::FromJson` is not called from GetElement/GetModel.
> 
> [Yes, it is](https://github.com/iTwin/imodel-native/blob/137a73e655d7a1336f181a89c3f93a216b3143b7/iModelJsNodeAddon/JsInteropDgnDb.cpp#L317).
> 
> [This caller](https://github.com/iTwin/itwinjs-core/blob/b99c65cdddf8b8e4975cff218c42b813864ab48e/core/backend/src/IModelDb.ts#L1578) expects any error thrown by the addon to have an `errorNumber`.

I thought you were talking about `DgnElements::GetElement`  (that doesn't/can't use JSON and would be bad to throw). 

If you attempt to load an element/model from JavaScript via Code and you supply an invalid scope, you will now get an exception that tells you that's what went wrong. I guess i should put some value (NotFound?) for errorNumber.

**Comment by pmconne** [comment]
> If you attempt to load an element/model from JavaScript via Code and you supply an invalid scope, you will now get an exception

That is why I'm concerned. We know that iModels can exist that contain elements with invalid scopes, and even after this PR it will still be possible to create new ones (by, e.g., deleting the scope element). So, it will be possible, for example, for an app to query the iIModel to obtain an invalid code, then attempt to obtain the properties of the element with that code and either get back no properties or produce an exception, depending on which API it calls.

Other ways to create invalid code scopes, I think, include running the transformer and omitting the code scope element, and connectors using the native API.

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp (hunk 3 above)
Comment (kabentley) [reply iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp:585]: > Querying for and loading entire element vs just its federation Guid;
This is rare and not worth worrying about.

>If caller supplies a valid but non-existent element Id, attempting fruitlessly to look up an element that has that Id as its federation Guid?
`scopeFederationGuid.FromString` will fail first.

**Comment by kabentley** [comment]
> That is why I'm concerned. We know that iModels can exist that contain elements with invalid scopes, and even after this PR it will still be possible to create new ones (by, e.g., deleting the scope element). So, it will be possible, for example, for an app to query the iIModel to obtain an invalid code, then attempt to obtain the properties of the element with that code and either get back no properties or produce an exception, depending on which API it calls.
> 
> Other ways to create invalid code scopes, I think, include running the transformer and omitting the code scope element, and connectors using the native API.

That's what I thought you were talking about originally. Nothing calls DgnCode::FromJson using data read from an iModel. Do I misunderstand your concern?

File: iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp (hunk 1 above)
Comment (kabentley) [reply iModelCore/iModelPlatform/DgnCore/CodeSpec.cpp]: i added it to `ElementExits`

**Comment by kabentley** [comment]
> IMO validation should happen on insert/update. It should not happen when constructing a code from JSON for other purposes.
I agree. Done.

**Review by pmconne — APPROVED** [review]

**Comment by pmconne** [comment]
@kabentley what's this waiting on?

**Pull request merged**